### PR TITLE
 Fix ICE involving mut references

### LINKED ISSUE #61623: rustc panic due to reference passed instead of mut reference
The following code causes a panic in rustc nightly and stable 2018 Edition:

```rust

use std::collections::HashMap;
use std::marker::PhantomData;
use std::str::FromStr;

struct Value<'a, T : FromStr + ToString + Clone, P : HasTree> {
    subkey : String,
    parent : &'a mut P,
    default : T,
    phantom : PhantomData<T>
}

pub trait HasTree {
    fn get_tree(&mut self) -> &mut Tree;
}

impl<'a, T : FromStr + ToString + Clone, P : HasTree> Value<'a, T, P> {
    fn from (subkey : String, parent : &'a mut P, default : T) -> Value<'a, T, P> {
        Value {
            subkey : subkey,
            parent : parent,
            default : default,
            phantom : PhantomData {}
        }
    }

    fn get(&self) -> T {
        let default = || { self.default.clone() };
        self.parent.get_tree().enter(&self.subkey).map_or_else(default, |subkey|
            subkey.data.as_ref().map_or_else(default, |str|
            str.parse::<T>().ok().unwrap_or_else(default)))
    }
}


#[derive(Clone)]
pub struct Tree {
    branches : HashMap<String, Tree>,
    pub data : Option<String>
}

impl Tree {
    pub fn new() -> Tree {
        Tree {
            branches : HashMap::new(),
            data : None
        }
    }

    fn enter(&self, s : &String) -> Option<&Tree> {
        self.branches.get(s)
    }
}
```

I do not expect this to compile, as get_tree requires a mut reference, which is not provided by get. When changing get_tree to `fn get_tree(&self) -> &Tree;`, the code compiles properly.

<details>
<summary>Backtrace</summary>
<pre>
   Compiling playground v0.0.1 (/playground)
thread 'rustc' panicked at 'no entry found for key', src/libcore/option.rs:1036:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:212
   6: rustc::util::common::panic_hook
   7: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:479
   8: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:382
   9: rust_begin_unwind
             at src/libstd/panicking.rs:309
  10: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  11: core::option::expect_failed
             at src/libcore/option.rs:1036
  12: rustc_mir::borrow_check::path_utils::each_borrow_involving_path
  13: rustc_mir::borrow_check::MirBorrowckCtxt::access_place
  14: <rustc_mir::borrow_check::MirBorrowckCtxt as rustc_mir::dataflow::DataflowResultsConsumer>::visit_statement_entry
  15: rustc_mir::borrow_check::do_mir_borrowck
  16: rustc::ty::context::GlobalCtxt::enter_local
  17: rustc_mir::borrow_check::mir_borrowck
  18: rustc::ty::query::__query_compute::mir_borrowck
  19: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_borrowck>::compute
  20: rustc::dep_graph::graph::DepGraph::with_task_impl
  21: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  22: rustc::ty::<impl rustc::ty::context::TyCtxt>::par_body_owners
  23: rustc::util::common::time
  24: rustc_interface::passes::analysis
  25: rustc::ty::query::__query_compute::analysis
  26: rustc::dep_graph::graph::DepGraph::with_task_impl
  27: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  28: rustc::ty::context::tls::enter_global
  29: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  30: rustc_interface::passes::create_global_ctxt::{{closure}}
  31: rustc_interface::interface::run_compiler_in_existing_thread_pool
  32: std::thread::local::LocalKey<T>::with
  33: scoped_tls::ScopedKey<T>::set
  34: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
query stack during panic:
#0 [mir_borrowck] processing `Value::<'a, T, P>::get`
#1 [analysis] running analysis passes on this crate
end of query stack

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.37.0-nightly (5eeb567a2 2019-06-06) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
</pre>

</details>

### LINKED ISSUE #61751: ICE: 'no entry found for key', src/libcore/option.rs:1034:5
ICE on both `1.35.0 stable` and `nightly 2019-06-10`

code [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5aa94087c049a1cb5da0ca45a4de52e7)
```rust
fn main() {
    let mut results = Vec::new();
    let immut_view = &results;
    std::thread::spawn(move || {
        std::thread::spawn(|| {
            immut_view.len();
        });
        for i in 0..5 {
            immut_view.push(i);
        }
    });
}
```

<details>
<summary> backtrace: cargo run +stable build --verbose</summary>

```
       Fresh lazy_static v1.3.0
       Fresh cfg-if v0.1.9
       Fresh smallvec v0.6.10
       Fresh glob v0.3.0
       Fresh crossbeam-utils v0.6.5
       Fresh crossbeam-channel v0.3.8
   Compiling scratch v0.1.0 (/home/alex/rust_projects/scratch)
     Running `sccache rustc --edition=2018 --crate-name scratch src/main.rs --color always --crate-type bin --emit=dep-info,link -C debuginfo=2 -C metadata=9f348a68eb7db5d8 -C extra-filename=-9f348a68eb7db5d8 --out-dir /home/alex/rust_projects/scratch/target/debug/deps -C incremental=/home/alex/rust_projects/scratch/target/debug/incremental -L dependency=/home/alex/rust_projects/scratch/target/debug/deps --extern crossbeam_channel=/home/alex/rust_projects/scratch/target/debug/deps/libcrossbeam_channel-c52d170e41357300.rlib --extern crossbeam_utils=/home/alex/rust_projects/scratch/target/debug/deps/libcrossbeam_utils-2ff7eb169679423f.rlib --extern glob=/home/alex/rust_projects/scratch/target/debug/deps/libglob-df35dc77c4d33d55.rlib`
thread 'rustc' panicked at 'no entry found for key', src/libcore/option.rs:1034:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   6: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   7: rust_begin_unwind
             at src/libstd/panicking.rs:308
   8: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   9: core::option::expect_failed
             at src/libcore/option.rs:1034
  10: rustc_mir::borrow_check::path_utils::each_borrow_involving_path
  11: rustc_mir::borrow_check::MirBorrowckCtxt::access_place
  12: <rustc_mir::borrow_check::MirBorrowckCtxt as rustc_mir::dataflow::DataflowResultsConsumer>::visit_statement_entry
  13: rustc_mir::borrow_check::do_mir_borrowck
  14: rustc::ty::context::GlobalCtxt::enter_local
  15: rustc_mir::borrow_check::mir_borrowck
  16: rustc::ty::query::__query_compute::mir_borrowck
  17: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_borrowck>::compute
  18: rustc::dep_graph::graph::DepGraph::with_task_impl
  19: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  20: rustc_mir::borrow_check::nll::type_check::TypeChecker::check_rvalue
  21: rustc_mir::borrow_check::nll::type_check::TypeChecker::typeck_mir
  22: rustc_mir::borrow_check::nll::type_check::type_check
  23: rustc_mir::borrow_check::nll::compute_regions
  24: rustc_mir::borrow_check::do_mir_borrowck
  25: rustc::ty::context::GlobalCtxt::enter_local
  26: rustc_mir::borrow_check::mir_borrowck
  27: rustc::ty::query::__query_compute::mir_borrowck
  28: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_borrowck>::compute
  29: rustc::dep_graph::graph::DepGraph::with_task_impl
  30: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  31: rustc::ty::<impl rustc::ty::context::TyCtxt>::par_body_owners
  32: rustc::util::common::time
  33: rustc_interface::passes::analysis
  34: rustc::ty::query::__query_compute::analysis
  35: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::analysis>::compute
  36: rustc::dep_graph::graph::DepGraph::with_task_impl
  37: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  38: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  39: rustc_interface::passes::create_global_ctxt::{{closure}}
  40: rustc_interface::interface::run_compiler_in_existing_thread_pool
  41: std::thread::local::LocalKey<T>::with
  42: scoped_tls::ScopedKey<T>::set
  43: syntax::with_globals
query stack during panic:
#0 [mir_borrowck] processing `main::{{closure}}#0`
#1 [mir_borrowck] processing `main`
#2 [analysis] running analysis passes on this crate
end of query stack

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.35.0 (3c235d560 2019-05-20) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `scratch`.

Caused by:
  process didn't exit successfully: `sccache rustc --edition=2018 --crate-name scratch src/main.rs --color always --crate-type bin --emit=dep-info,link -C debuginfo=2 -C metadata=9f348a68eb7db5d8 -C extra-filename=-9f348a68eb7db5d8 --out-dir /home/alex/rust_projects/scratch/target/debug/deps -C incremental=/home/alex/rust_projects/scratch/target/debug/incremental -L dependency=/home/alex/rust_projects/scratch/target/debug/deps --extern crossbeam_channel=/home/alex/rust_projects/scratch/target/debug/deps/libcrossbeam_channel-c52d170e41357300.rlib --extern crossbeam_utils=/home/alex/rust_projects/scratch/target/debug/deps/libcrossbeam_utils-2ff7eb169679423f.rlib --extern glob=/home/alex/rust_projects/scratch/target/debug/deps/libglob-df35dc77c4d33d55.rlib` (exit code: 101)

```
</details>

### LINKED ISSUE #61944: `internal compiler error: unexpected panic` on 7 line double ref program
I compiled the following code expecting some sort of error message. Instead, the compiler panicked unexpectedly. ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c07140d75f2671cc7e995b334f2722b2))

```rust
fn main() {
    let x = String::from("foo");
    let x_ref = &x;
    let x_ref_ref = &x_ref;
    x_ref.push_str("bar");
    println!("{}", x_ref_ref);
}
```

`cargo run` errors:

```
Compiling hello v0.1.0 (/home/vagrant/hello)
thread 'rustc' panicked at 'no entry found for key', src/libcore/option.rs:1034:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.35.0 (3c235d560 2019-05-20) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `hello`.

To learn more, run the command again with --verbose.
```

`RUST_BACKTRACE=1 cargo run --verbose` errors: 

```
   Compiling hello v0.1.0 (/home/vagrant/hello)
     Running `rustc --edition=2018 --crate-name hello src/main.rs --color always --crate-type bin --emit=dep-info,link -C debuginfo=2 -C metadata=3088e465bdea9221 -C extra-filename=-3088e465bdea9221 --out-dir /home/vagrant/hello/ta
rget/debug/deps -C incremental=/home/vagrant/hello/target/debug/incremental -L dependency=/home/vagrant/hello/target/debug/deps`
thread 'rustc' panicked at 'no entry found for key', src/libcore/option.rs:1034:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print                                                                                                                                                                                        [0/372]
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   6: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   7: rust_begin_unwind
             at src/libstd/panicking.rs:308
   8: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   9: core::option::expect_failed
             at src/libcore/option.rs:1034
  10: rustc_mir::borrow_check::path_utils::each_borrow_involving_path
  11: rustc_mir::borrow_check::MirBorrowckCtxt::access_place
  12: <rustc_mir::borrow_check::MirBorrowckCtxt as rustc_mir::dataflow::DataflowResultsConsumer>::visit_statement_entry
  13: rustc_mir::borrow_check::do_mir_borrowck
  14: rustc::ty::context::GlobalCtxt::enter_local
  15: rustc_mir::borrow_check::mir_borrowck
  16: rustc::ty::query::__query_compute::mir_borrowck
  17: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_borrowck>::compute
  18: rustc::dep_graph::graph::DepGraph::with_task_impl
  19: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  20: rustc::ty::<impl rustc::ty::context::TyCtxt>::par_body_owners
  21: rustc::util::common::time
  22: rustc_interface::passes::analysis
  23: rustc::ty::query::__query_compute::analysis
  24: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::analysis>::compute
  25: rustc::dep_graph::graph::DepGraph::with_task_impl
  26: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  27: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  28: rustc_interface::passes::create_global_ctxt::{{closure}}
  29: rustc_interface::interface::run_compiler_in_existing_thread_pool
  30: std::thread::local::LocalKey<T>::with
  31: scoped_tls::ScopedKey<T>::set
  32: syntax::with_globals
query stack during panic:
#0 [mir_borrowck] processing `main`
#1 [analysis] running analysis passes on this crate
end of query stack

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.35.0 (3c235d560 2019-05-20) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `hello`.

Caused by:
  process didn't exit successfully: `rustc --edition=2018 --crate-name hello src/main.rs --color always --crate-type bin --emit=dep-info,link -C debuginfo=2 -C metadata=3088e465bdea9221 -C extra-filename=-3088e465bdea9221 --out-dir /home/vagrant/hello/target/debug/deps -C incremental=/home/vagrant/hello/target/debug/incremental -L dependency=/home/vagrant/hello/target/debug/deps` (exit code: 101)
```

`rustc --version --verbose`: 

```
rustc 1.35.0 (3c235d560 2019-05-20)
binary: rustc
commit-hash: 3c235d5600393dfe6c36eeed34042efad8d4f26e
commit-date: 2019-05-20
host: x86_64-unknown-linux-gnu
release: 1.35.0
LLVM version: 8.0
```

### PR DESCRIPTION
**Pull request by estebank**
Fix #61623, fix #61944, fix #61751.

### CODE REVIEW

**Comment by estebank** [comment]
I need to investigate if the changes in the second commit are spurious to my local env.

File: src/librustc_mir/borrow_check/mod.rs [hunk 1]
```diff
@@ -1070,7 +1070,7 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
 
                 (Reservation(WriteKind::MutableBorrow(bk)), BorrowKind::Shallow)
                 | (Reservation(WriteKind::MutableBorrow(bk)), BorrowKind::Shared) if {
-                    tcx.migrate_borrowck()
+                    tcx.migrate_borrowck() && this.borrow_set.location_map.get(&location).is_some()
```
Comment (matthewjasper) [inline src/librustc_mir/borrow_check/mod.rs]: This can use `contains_key`

**Review by matthewjasper — COMMENTED** [review]
r=me with nit addressed.
r? @matthewjasper

**Comment by pnkfelix** [comment]
Discussed at T-compiler meeting; accepting  for beta-backport.

**Pull request merged**